tools/pubsys/src/aws/ami/mod.rs

```diff
@@ -372,20 +418,63 @@ async fn _run(args: &Args, ami_args: &AmiArgs) -> Result<HashMap<String, Image>>
     Ok(amis)
 }
 
+#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq, Hash)]
+pub(crate) struct LaunchPermissionDef {
+    /// The name of the group
+    pub group: Option<String>,
+
+    /// The Amazon Web Services account ID
+    pub user_id: Option<String>,
+
+    /// The ARN of an organization
+    pub organization_arn: Option<String>,
+
+    /// The ARN of an organizational unit
+    pub organizational_unit_arn: Option<String>,
+}
+
+impl From<LaunchPermission> for LaunchPermissionDef {
+    fn from(launch_permission: LaunchPermission) -> Self {
+        Self {
+            group: launch_permission
+                .group()
+                .map(|group| group.as_str().to_owned()),
+            user_id: launch_permission
+                .user_id()
```

Comment:
Supposing you do make this an enum, you could rewrite this pretty succinctly with a match:

```rust
let LaunchPermission { group, user_id, organization_arn, organizational_unit_arn } = launch_permission;
    match (group, user_id, organization_arn, organizational_unit_arn) {
        (Some(group), None, None, None) => LaunchPermissionDef::Group(group),
        (None, Some(user_id), None, None) => LaunchPermissionDef::UserId(user_id),
        (None, None, Some(organization_arn), None) => LaunchPermissionDef::OrganizationArn(organization_arn),
        (None, None, None, Some(organizational_unit_arn)) => LaunchPermissionDef::OrganizationalUnitArn(organizational_unit_arn),
        _ => // error!
    };
```


You would have to change this from `impl From` to `impl TryFrom` to handle the error case of all of the elements being `None` though.